cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java

```diff
@@ -267,16 +267,28 @@ protected Program createProgram(CConfiguration cConf, ProgramRunner programRunne
                                   ProgramDescriptor programDescriptor, ArtifactDetail artifactDetail,
                                   final File tempDir) throws IOException {
 
+    // TODO: (CDAP-19150) remove Location usage from ArtifactDetail
     Location programJarLocation = artifactDetail.getDescriptor().getLocation();
     ClassLoaderFolder classLoaderFolder;
     try {
       // If the program jar is not a directory, take a snapshot of the jar file to avoid mutation.
       if (!programJarLocation.isDirectory()) {
         File targetFile = new File(tempDir, "program.jar");
-        try {
-          programJarLocation = Locations.toLocation(Locations.linkOrCopyOverwrite(programJarLocation, targetFile));
-        } catch (FileAlreadyExistsException ex) {
-          LOG.warn("Program file {} already exists and can not be replaced.", targetFile.getAbsolutePath());
+        if (!programJarLocation.exists()) {
+          // This hack is here because the program jar won't actually exist in a tethered run
+          // TODO: (CDAP-19150) remove hack once Location is removed form ArtifactDetail
```

Comment:
Can we do this if it's a tethered run? https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java#L161